include/coap/coap_dtls.h

```diff
@@ -43,14 +100,58 @@ int coap_dtls_get_log_level(void);
 void *
 coap_dtls_new_context(struct coap_context_t *coap_context);
 
+/**
+ * Set the dtls context's default server PSK hint and/or key.
+ * This does the PSK specifics for coap_dtls_new_context()
+ *
+ * @param ctx The CoAP context.
+ * @param hint    The default PSK server hint sent to a client. If NULL, PSK
+ *                authentication is disabled. Empty string is a valid hint.
+ * @param key     The default PSK key. If NULL, PSK authentication will fail.
+ * @param key_len The default PSK key's lenght. If 0, PSK authentication will
+ *                fail.
+ *
+ * @return 1 if successful, else 0
+ */
+
+int coap_dtls_context_set_psk(struct coap_context_t *ctx, const char *hint,
+                           const uint8_t *key, size_t key_len );
+
+/**
+ * Set the dtls context's default server PKI information.
+ * This does the PKI specifics for coap_dtls_new_context()
+ * The Callback is called to set up the appropriate information.
+ *
+ * @param ctx The CoAP context.
+ * @param setup_callback The callback which is TLS library support dependent
+ * @param setup_data     If NULL, PKI authentication will fail. Certificate
+ *                       information required.
+ *
+ * @return 1 if successful, else 0
+ */
+
+int coap_dtls_context_set_pki(struct coap_context_t *ctx,
```

Comment:
Typically in libcoap when a function takes a callback and a data argument, the callback is required and the data is optional and opaque (for use in the user-supplied callback only). Here the convention is reversed, the callback is optional and the data is mandatory, and used by libcoap. One suggestion would be to put the callback inside the coap_dtls_pki_t structure to avoid the confusion and underline the optional nature of the callback.

Reply:
Do you want me to make this change and push it?

Reply:
Callback now inside coap_dtls_pki_t structure and added to this pull request